cluster-test/test-orchestrator/test_yaml_parser.go

```diff
@@ -0,0 +1,308 @@
+			err = errors.New(fmt.Sprint(r))
+		}
+	}()
+
+	return parseScenarios(bts), nil
+}
+
+func parseScenarios(bts []byte) []*Scenario {
+	testYaml := &testYaml{}
+	err := yaml.Unmarshal([]byte(bts), testYaml)
+	if err != nil {
+		panic("failed to unmarshal scenario yaml")
+	}
+
+	return extractScenarios(testYaml)
+}
+
+// extractScenarios returns a scenario for every element in the runs list.
+func extractScenarios(runs *testYaml) []*Scenario {
+	var result []*Scenario
+	for _, scenarioData := range runs.Scenarios {
+		for i := 1; i < len(scenarioData.Runs); i++ {
+			s := extractScenario(scenarioData, i)
+			result = append(result, s)
+		}
+	}
+
+	return result
+}
+
+// extractScenario returns a scenario given the index of a specific run.
+func extractScenario(data *scenarioYaml, runIx int) *Scenario {
+	varsData := data.Runs[0]
+	runData := data.Runs[runIx]
+	defer wrapPanicf("Failed to parse scenario '%s'", data.Name)
+	defer wrapPanicf("in run %d, [%v] = [%v]", runIx, strings.Join(varsData, ", "), strings.Join(runData, ", "))
+
+	if len(varsData) != len(runData) {
+		msg := fmt.Sprintf("var count of run %v should match var count of %v", runData, varsData)
+		panic(msg)
+	}
+
+	// don't find and replace on name
+	name := data.Name
+	desc := replace(data.Desc, varsData, runData)
+	sizeStr := replace(data.Size, varsData, runData)
+
+	// extract size
+	size, err := strconv.Atoi(sizeStr)
+	if err != nil {
+		panic("size convert: " + err.Error())
+	}
+
+	// extract script
+	labels, cmds := extractScript(data.Script, varsData, runData)
+	script := parseScript(labels, cmds)
+
+	// extract Measure
+	var measureStrs []string
+	for _, measureStr := range data.Measure {
+		measureStrs = append(measureStrs, replace(measureStr, varsData, runData))
+	}
+	measure := parseMeasure(measureStrs)
+
+	return &Scenario{
+		Name:    name,
+		Desc:    desc,
+		Size:    size,
+		Script:  script,
+		Measure: measure,
+	}
+}
+
+func extractScript(script []map[string]string, varsData, runData []string) (labels, cmds []string) {
+	for _, cmdData := range script {
+		if len(cmdData) != 1 {
+			msg := fmt.Sprintf("\"%v\" is not a valid command, should contain exactly one entry", cmdData)
+			panic(msg)
+		}
+
+		for label, cmd := range cmdData {
+			labels = append(labels, replace(label, varsData, runData))
+			cmds = append(cmds, replace(cmd, varsData, runData))
+		}
+	}
+	return labels, cmds
+}
+
+func parseScript(labels, cmdStrs []string) []*Command {
+	defer wrapPanicf("in parse script")
+	var cmds []*Command
+	for i := range labels {
+		cmd := parseCommand(labels[i], cmdStrs[i])
+		cmds = append(cmds, cmd)
+	}
+
+	return cmds
+}
+
+func parseCommand(label, cmdString string) *Command {
+	defer wrapPanicf("in parse command '%s: %s'", label, cmdString)
+	split := split(cmdString)
+	if len(split) == 0 {
+		panic("empty command")
+	}
+
+	return &Command{
+		Label: label,
+		Cmd:   split[0],
+		Args:  split[1:],
+	}
+}
+
+func parseMeasure(msData []string) []*Measurement {
```

Comment:
Not really, it's parsing the measure field in the yaml. Have no problem changing it though